File: JS/Tree/InorderTraversalIter.js
// Definition for a  binary tree node
//			function TreeNode(data){
//				this.data = data
//				this.left = null
//				this.right = null
//			}

module.exports = {
	//param root : root node of tree
	//return a array of integers
	inorderTraversal : function(root){
        var res = [], stack = [];
        stack.push(root);
        while(stack.length > 0){
            root = root.left;
            while(root == null && stack.length > 0){
                root = stack.pop();
                res.push(root.data);
                root = root.right;
            }
            if(root != null)stack.push(root);
        }
        return res;
	}
};

File: Java/Trees/InorderTraversalIter.java
/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	public ArrayList<Integer> inorderTraversal(TreeNode a) {
	    ArrayList<Integer> res = new ArrayList<Integer>();
	    if(a == null)return res;
	    Stack<TreeNode> stack = new Stack<>();
	    stack.push(a);
	    while(true){
	        a = a.left;
	        while(a == null && !stack.isEmpty()){
	            a = stack.pop();
	            res.add(a.val);
	            a = a.right;
	        }
	        if(a == null)break;
	        stack.push(a);
	    }
	    return res;
	}

}
